Source/WebCore/dom/TrustedType.cpp

```diff
@@ -168,4 +168,40 @@ ExceptionOr<String> trustedTypeCompliantString(TrustedType expectedType, ScriptE
     return stringValue;
 }
 
+// https://w3c.github.io/trusted-types/dist/spec/#require-trusted-types-for-pre-navigation-check
+ExceptionOr<URL> requireTrustedTypesForPreNavigationCheckPasses(ScriptExecutionContext& scriptExecutionContext, const String& urlString)
+{
+    auto sinkGroup = "script"_s;
+    auto sink = "Location href"_s;
+    auto expectedType = TrustedType::TrustedScript;
+
+    auto contentSecurityPolicy = scriptExecutionContext.contentSecurityPolicy();
+
+    auto requireTrustedTypes = contentSecurityPolicy && contentSecurityPolicy->requireTrustedTypesForSinkGroup(sinkGroup);
+
+    if (!requireTrustedTypes || !scriptExecutionContext.settingsValues().trustedTypesEnabled)
+        return URL(urlString);
+
+    const int javascriptSchemeLength = sizeof("javascript:") - 1;
+    auto decodedURLString = PAL::decodeURLEscapeSequences(urlString);
+    auto scriptSource = decodedURLString.substring(javascriptSchemeLength);
+
+    auto convertedScriptSource = processValueWithDefaultPolicy(scriptExecutionContext, expectedType, scriptSource, sink);
+    if (std::holds_alternative<Exception>(convertedScriptSource))
+        return WTFMove(std::get<Exception>(convertedScriptSource));
+
+    if (std::holds_alternative<std::monostate>(convertedScriptSource)) {
+        auto allowMissingTrustedTypes = contentSecurityPolicy->allowMissingTrustedTypesForSinkGroup(trustedTypeToString(expectedType), sink, sinkGroup, scriptSource);
+
+        if (allowMissingTrustedTypes)
+            return URL(urlString);
+
+        return Exception { ExceptionCode::TypeError, makeString("This assignment requires a ", trustedTypeToString(expectedType)) };
```

Comment:
I would tend to do `if (!allowMissingTrustedTypes) return Exception...`

Reply:
I was between the two so will swap it